### DISCUSSION TITLE
How do I use atomFamily with a setter and an async getter?

### DISCUSSION THREAD

**Discussion by jessepinho** [pmndrs/jotai#2162] [Q&A]
I'm building a React Native app whose basic data model is similar to Instagram's — i.e., a bunch of posts. There are two needs that I'm trying to solve with Jotai.

## Need number 1: be able to load posts asynchronously without the component knowing about it

I'd like to create a `postsAtomFamily` that I can use in my `SinglePostScreen` like so:

```TSX
export default function SinglePostScreen({ id }: { id: string }) {
  const post = useAtomValue(postsAtomFamily(id))

  return <Text>{post.title}</Text>
}
```

The idea is, even if I've never loaded the post before, I can just grab it from `postsAtomFamily`, and the atom will take care of the requesting logic without the component having to worry about it. React Suspense will simply hide my component until the data is ready.

Then, if the user happens to revisit the `SinglePostScreen` for the post with the exact same post ID, `postsAtomFamily` should just return the post from cache rather than requesting a new one.

## Need number 2: be able to update the `postsAtomFamily` "cache" from other places in the app

Separately, I have a `HomeFeedScreen` component, where I'd like to do this:

```TSX
export default function HomeFeedScreen() {
  const posts = useAtomValue(homeFeedAtom)

  return (
    <>
      {posts.map(post => (
        <Text key={post.id}>{post.title}</Text>
      ))}
    </>
  )
}
```

To render the home feed, we have to make a different API call than the one we make for the `SinglePostScreen`. For the `SinglePostScreen`, we're using `postsAtomFamily` to make an API call for a single post. But for the home feed, we need to use `homeFeedAtom` to request a whole array of posts rather than just a single post.

So what I'd like to do is this: when I request the whole array of posts for the home screen, I want to _also_ update the `postsAtomFamily` cache with all the posts we just received for the home feed. That way, no matter where we view a post (or series of posts) in the app, they're all being cached in the same `postsAtomFamily` cache. And if a user navigates from the home feed to the `SinglePostScreen`, the post that was loaded in the home feed won't need to be re-requested from the server — it'll already be in `postsAtomFamily`.

---

The only way I could figure out how to solve this was to create a `postsAtomFamily` like this:

```TypeScript
const postsCacheAtom = atom<Map<string, Post>>(new Map())

const postsAtomFamily = atomFamily((id: string) =>
  atom(
    async get => {
      // If somewhere else in the app, we've manually set the post with this ID, it'll be in the posts cache atom defined above. So let's just retrieve and return it.
      const postsCache = get(postsCacheAtom)
      if (postsCache.has(id)) return postsCache.get(id)

      // Otherwise, call out to the API to retrieve the post. (Then, on future calls, this post will be retrieved via the internal `atomFamily` cache map, rather than `postsCacheAtom` above.)
      const myApi = get(apiAtom)
      return await myApi.getPostById(id)
    },
    (get, set, update: Post) => {
      // If e.g., the home feed calls `set(postsAtomFamily('some-id'), somePost)`, we'll add it to the posts cache here.
      const currentCache = get(postsCacheAtom)
      const newCache = new Map(currentCache)
      newCache.set(update.id, update)
      set(postsCacheAtom, newCache)
    },
  ),
)

export default postsAtomFamily
```

This... works? But it seems super weird. There are basically two caches: one is the internal cache created by `atomFamily`; the other is the `postsCacheAtom` I created myself. Which of the two caches gets used depends on whether you're trying to retrieve a post via the `SinglePostScreen` or via the home screen. (The former would get cached in `atomFamily`'s internal cache; the latter gets cached in the `postsCacheAtom`, when the home screen manually sets it after requesting the home feed from the API.)

Speaking of the home feed, here's what that looks like:

```TSX
const homeFeedAtom = atom<Post[]>([])

const loadHomeFeedAtom = atom(null, async (get, set) => {
  const myApi = get(apiAtom)
  return await myApi
    .getLatestPosts()
    .then(posts => {
      // Add all posts to the cache
      posts.forEach(post => {
        set(postsAtomFamily(post.id), post)
      })
      // Also set them on the `homeFeedAtom` so we can use them on the home screen. Still need to figure out how to save _references_ to them, rather than the posts themselves (i.e., so that `postsAtomFamily` is our centralized source of truth for posts).
      set(homeFeedAtom, posts)
    })
})
```

I then have to call `loadHomeFeedAtom()` in an effect on the home screen, since just using the getter won't give me access to `set()`. This also feels clumsy — I want to just consume the home feed atom without having to call a function that hits the API.

I feel like there's something wrong with my mental model for how to use this. Can you help guide me to achieve what I'm going for? I freaking love the library but I can't figure out how to use it for my use case!

**Comment by dai-shi** [comment]
atomFamily + atomWithDefault would fit better.

**Reply by jessepinho** [reply]
Ah, thank you! That would definitely solve need number 1. 

For need number 2 though, is there a way for me to populate the atom family while requesting the full home feed, without having to explicitly use an "action" atom? i.e., can I do this from the home screen:

```TSX
export default function HomeFeedScreen() {
  const homeFeed = useAtomValue(homeFeedAtom)
}
```

And then in the background, `homeFeedAtom` starts requesting the posts for the home feed, _**AND**_ populates `postsAtomFamily` when finished?

As far as I understand, it's not possible to set the value of one atom from the getter of another atom, but just thought I'd ask :) Thanks!

**Reply by dai-shi** [reply]
In such a case, https://jotai.org/docs/guides/atoms-in-atom might fit better than `atomFamily`.
